### ISSUE TITLE
Mismatch between unexpanded and expanded ASTs

### ISSUE THREAD

**Issue by Akeboshiwind** [immunant/c2rust#224]
Hey, I ran the following command on the [rustybox](https://github.com/samuela/rustybox) codebase on [master](https://github.com/samuela/rustybox/tree/b7d7f01bdff8a85dfb7dcf984999b82a8ba0757a):
```bash
$ RUST_BACKTRACE=1 c2rust refactor --cargo -r print select target 'crate; desc(match_expr(2));'
```

And got the following error:
```
thread 'main' panicked at 'mismatch between unexpanded and expanded ASTs', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/c2rust-refactor-0.14.0/src/collapse/mac_table.rs:404:23
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:84
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1024
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1428
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:193
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: rustc_driver::report_ice
  11: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at /rustc/6d77e45f01079fe3d40180b3e256e414ab379f63/src/liballoc/boxed.rs:983
  12: proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter::{{closure}}::{{closure}}
             at /rustc/6d77e45f01079fe3d40180b3e256e414ab379f63/src/libproc_macro/bridge/client.rs:305
  13: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
  14: std::panicking::begin_panic
  15: <syntax::ast::Item as c2rust_refactor::collapse::mac_table::CollectMacros>::collect_macros
  16: c2rust_refactor::collapse::mac_table::collect_macros_seq
  17: <syntax::ast::Item as c2rust_refactor::collapse::mac_table::CollectMacros>::collect_macros
  18: c2rust_refactor::collapse::mac_table::collect_macros_seq
  19: <syntax::ast::Item as c2rust_refactor::collapse::mac_table::CollectMacros>::collect_macros
  20: c2rust_refactor::collapse::mac_table::collect_macros_seq
  21: c2rust_refactor::collapse::mac_table::collect_macro_invocations
  22: c2rust_refactor::collapse::CollapseInfo::collect
  23: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  24: <c2rust_refactor::command::DriverCommand<F> as c2rust_refactor::command::Command>::run
  25: c2rust_refactor::command::RefactorState::run
  26: std::thread::local::LocalKey<T>::with
  27: scoped_tls::ScopedKey<T>::set
  28: syntax::with_globals
  29: c2rust_refactor::driver::run_refactoring
  30: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  31: std::panicking::try::do_call
  32: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  33: rustc_driver::catch_fatal_errors
  34: c2rust_refactor::lib_main
  35: c2rust_refactor::main
  36: std::rt::lang_start::{{closure}}
  37: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  38: std::panicking::try::do_call
             at src/libstd/panicking.rs:292
  39: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  40: std::panicking::try
             at src/libstd/panicking.rs:270
  41: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  42: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  43: main
  44: __libc_start_main
  45: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

error: Unrecognized option: 'cargo'
```

Versions:
```bash
$ c2rust refactor --version
c2rust-refactor 0.9.0
```

I'm running this all in a docker container on macos using the following Dockerfile:
```Dockerfile
FROM rust:1.40

WORKDIR /usr/src

# Install deps
RUN apt update && \
    apt install -y build-essential llvm-6.0 clang-6.0 libclang-6.0-dev cmake libssl-dev pkg-config python3

# Install required rust nightly version
RUN rustup install nightly-2019-12-05 && \
    rustup component add --toolchain nightly-2019-12-05 rustfmt && \
    rustup component add --toolchain nightly-2019-12-05 rustc-dev

# Install c2rust
RUN LLVM_CONFIG_PATH=/usr/lib/llvm-6.0/bin/llvm-config cargo +nightly-2019-12-05 install c2rust

# Install dev tools
RUN apt install -y vim
```

**Comment by rinon** [comment]
It looks like deriving `PartialEq` on a struct is breaking the refactoring tool's macro collapser. I'll look into this.

**Comment by rinon** [comment]
This should be fixed by 85f34587e5fe2430178e2cc8ae14952de1aef22a, could you try with the latest master version?

**Comment by Akeboshiwind** [comment]
It's been long enough that I can't even remember what the command I wanted originally did 😅.

But I've run the command I originally posted and I can confirm that I no longer get a panic 🎉
